tidy3d/components/source/time.py

```diff
@@ -162,7 +224,72 @@ def end_time(self) -> Optional[float]:
         # if not self.remove_dc_component:
         #     return None
 
-        return self.offset * self.twidth + END_TIME_FACTOR_GAUSSIAN * self.twidth
+        end_time = self.offset_time + END_TIME_FACTOR_GAUSSIAN * self.twidth
+
+        # for derivative Gaussian that contains two peaks, add time interval between them
+        if self.remove_dc_component and self.fwidth > self.freq0:
+            end_time += 2 * self._peak_time_shift
+        return end_time
+
+    def amp_freq(self, freq: float) -> complex:
+        """Complex-valued source spectrum in frequency domain."""
+        phase = np.exp(1j * self.phase + 1j * 2 * np.pi * (freq - self.freq0) * self.offset_time)
+        envelope = np.exp(-((freq - self.freq0) ** 2) / 2 / self.fwidth**2)
+        amp = 1j * self.amplitude / self.fwidth * phase * envelope
+        if not self.remove_dc_component:
+            return amp
+
+        # derivative of Gaussian when DC is removed
+        return freq * amp / (2 * np.pi * self.peak_frequency)
+
+    def _rel_amp_freq(self, freq: float) -> complex:
+        """Complex-valued source spectrum in frequency domain normalized by peak amplitude."""
+        return self.amp_freq(freq) / self._peak_freq_amp
+
+    @property
+    def peak_frequency(self) -> float:
+        """Frequency at which the source time dependence has its peak amplitude in the frequency domain."""
+        if not self.remove_dc_component:
+            return self.freq0
+        return 0.5 * (self.freq0 + np.sqrt(self.freq0**2 + 4 * self.fwidth**2))
+
+    @property
+    def _peak_freq_amp(self) -> complex:
+        """Peak amplitude in frequency domain"""
+        return self.amp_freq(self.peak_frequency)
+
+    @property
+    def _peak_time_amp(self) -> complex:
+        """Peak amplitude in time domain"""
+        return self.amp_time(self.peak_time)
+
+    def frequency_range_sigma(self, sigma: float = DEFAULT_SIGMA) -> FreqBound:
+        """Frequency range where the source amplitude is within ``exp(-sigma**2/2)`` of the peak amplitude."""
+        if not self.remove_dc_component:
+            return self.frequency_range(num_fwidth=sigma)
+
+        # With dc removed, we'll need to solve for the transcendental equation to find the frequency range
+        def equation_for_sigma_frequency(freq):
+            """computes A / A_p - exp(-sigma)"""
+            return np.abs(self._rel_amp_freq(freq)) - np.exp(-(sigma**2) / 2)
+
+        root_scalar = Brentq(raise_on_fail=False, epsilon=_ROOTS_TOL)
+        fmin_data = root_scalar(equation_for_sigma_frequency, 0, self.peak_frequency)
+        fmax_data = root_scalar(
+            equation_for_sigma_frequency,
+            self.peak_frequency,
+            self.peak_frequency
+            + self.fwidth
+            * (OFFSET_FWIDTH_FMAX + 2 * sigma**2),  # "100" to make sure that it flips sign
```

Comment:
`Offset` instead of "100" in the comment? Also, could you give the argument names in these invocations so it's clear what's what.